File: C++/33 Search in Rotated Sorted Array.cpp
class Solution
{
public:
    int search(vector<int> &nums, int target)
    {
        auto search = [&](vector<int> &nums, int target, int l, int r) -> int
        {
            while (l < r)
            {
                int mid = l + r >> 1;
                if (target <= nums[mid])
                {
                    r = mid;
                }
                else
                {
                    l = mid + 1;
                }
            }
            return l;
        };

        int n = nums.size();
        if (nums[0] < nums[n - 1])
        {
            int x = search(nums, target, 0, n - 1);
            if (nums[x] == target)
            {
                return x;
            }
        }
        else
        {
            int l = 0, r = n - 1;
            while (l < r)
            {
                int mid = l + r >> 1;
                if (nums[0] > nums[mid])
                {
                    r = mid;
                }
                else
                {
                    l = mid + 1;
                }
            }

            int idx = -1;
            if (target >= nums[0])
            {
                idx = search(nums, target, 0, l - 1);
            }
            else
            {
                idx = search(nums, target, l, n - 1);
            }
            if (nums[idx] == target)
            {
                return idx;
            }
        }

        return -1;
    }
};

File: Swift/33 Search in Rotated Sorted Array.swift
class Solution {
    func search(_ nums: [Int], _ target: Int) -> Int {
        var n = nums.count
        var k = n - 1

        func binary(_ left: Int, _ right: Int) -> Int {
            var l = left, r = right
            while l < r {
                var mid = (l + r) >> 1
                if target <= nums[mid] {
                    r = mid
                } else {
                    l = mid + 1
                }
            }
            return l
        }

        if nums[0] < nums[k] {
            let x = binary(0, k)
            if nums[x] == target {
                return x
            }
        } else {
            var l = 0, r = k
            while l < r {
                let mid = (l + r) / 2
                if nums[0] > nums[mid] {
                    r = mid
                } else {
                    l = mid + 1
                }
            }

            var idx: Int = -1
            if nums[0] <= target {
                idx = binary(0, l - 1)
            } else {
                idx = binary(l, k)
            }
            if nums[idx] == target {
                return idx
            }
        }
        return -1
    }
}
